hw/src/trezor.rs

```diff
@@ -0,0 +1,431 @@
+			Error::KeyNotFound => write!(f, "Key not found"),
+			Error::UserCancel => write!(f, "Operation has been cancelled"),
+			Error::BadMessageType => write!(f, "Bad Message Type in RPC call"),
+			Error::SerdeError(ref e) => write!(f, "Serde serialization error: {}", e),
+			Error::ClosedDevice(ref s) => write!(f, "Device is closed, needs PIN to perform operations: {}", s),
+		}
+	}
+}
+
+impl From<hidapi::HidError> for Error {
+	fn from(err: hidapi::HidError) -> Error {
+		Error::Usb(err)
+	}
+}
+
+impl From<protobuf::ProtobufError> for Error {
+	fn from(_: protobuf::ProtobufError) -> Error {
+		Error::Protocol(&"Could not read response from Trezor Device")
+	}
+}
+
+/// Ledger device manager.
+pub struct Manager {
+	usb: Arc<Mutex<hidapi::HidApi>>,
+	devices: Vec<Device>,
+	closed_devices: Vec<String>,
+	key_path: KeyPath,
+}
+
+#[derive(Debug)]
+struct Device {
+	path: String,
+	info: WalletInfo,
+}
+
+/// HID Version used for the Trezor device
+enum HidVersion {
+	V1,
+	V2,
+}
+
+impl Manager {
+	/// Create a new instance.
+	pub fn new(hidapi: Arc<Mutex<hidapi::HidApi>>) -> Manager {
+		Manager {
+			usb: hidapi,
+			devices: Vec::new(),
+			closed_devices: Vec::new(),
+			key_path: KeyPath::Ethereum,
+		}
+	}
+
+	/// Re-populate device list
+	pub fn update_devices(&mut self) -> Result<usize, Error> {
+		let mut usb = self.usb.lock();
+		usb.refresh_devices();
+		let devices = usb.devices();
+		let mut new_devices = Vec::new();
+		let mut closed_devices = Vec::new();
+		for usb_device in devices {
+			trace!("Checking device: {:?}", usb_device);
+			if usb_device.vendor_id != TREZOR_VID || !TREZOR_PIDS.contains(&usb_device.product_id) || usb_device.usage_page != 0xFF00 {
+				continue;
+			}
+			match self.read_device_info(&usb, &usb_device) {
+				Ok(device) => new_devices.push(device),
+				Err(Error::ClosedDevice(path)) => closed_devices.push(path.to_string()),
+				Err(e) => return Err(e),
+			}
+		}
+		let count = new_devices.len();
+		self.devices = new_devices;
+		self.closed_devices = closed_devices;
+		Ok(count)
+	}
+
+	fn read_device_info(&self, usb: &hidapi::HidApi, dev_info: &hidapi::HidDeviceInfo) -> Result<Device, Error> {
+		let handle = self.open_path(|| usb.open_path(&dev_info.path))?;
+		let manufacturer = dev_info.manufacturer_string.clone().unwrap_or("Unknown".to_owned());
+		let name = dev_info.product_string.clone().unwrap_or("Unknown".to_owned());
+		let serial = dev_info.serial_number.clone().unwrap_or("Unknown".to_owned());
+		match self.get_address(&handle) {
+			Ok(Some(addr)) => {
+				Ok(Device {
+					path: dev_info.path.clone(),
+					info: WalletInfo {
+						name: name,
+						manufacturer: manufacturer,
+						serial: serial,
+						address: addr,
+					},
+				})
+			}
+			Ok(None) => Err(Error::ClosedDevice(dev_info.path.clone())),
+			Err(e) => Err(e),
+		}
+	}
+
+	pub fn message(&self, message_type: String, device_path: Option<String>, message: Option<String>) -> Result<String, Error> {
```

Comment:
do any of these passed parameters actually have to be owned data?

Reply:
No, changing it!